feat(http source): Add support for specifying http method

r3b-fish opened the pull request:
Allow to configure http method to process input requests. 

Sometimes, we can't configure hooks with POST. We should use proxing or add extra configuration to ingress controllers.

jszwedko commented on src/sources/http.rs:
```diff
@@ -43,6 +43,8 @@ pub(super) struct SimpleHttpConfig {
     path: String,
     #[serde(default = "default_path_key")]
     path_key: String,
+    #[serde(default = "default_http_method")]
+    http_method: String,
```
I think it'd be best to define an enum for this to deserialize into so that users get a configuration error if they use an unsupported `http_method`.

jszwedko commented on website/cue/reference/components/sources/http.cue:
```diff
@@ -128,6 +128,15 @@ components: sources: http: {
 				examples: ["vector_http_path"]
 			}
 		}
+		http_method: {
+			common:      false
+			description: "Specifies the action of the HTTP request."
+			required:    false
+			type: string: {
+				default: "POST"
+				examples: ["HEAD", "GET", "PUT", "POST", "PATCH", "DELETE"]
+			}
+		}
```
I think this should be an `enum` to show users the only valid methods. See

https://github.com/vectordotdev/vector/blob/97e39a9ac1f61f02c61e549009ac1fe2c9ff6d17/website/cue/reference/components/sources/aws_s3.cue#L59-L72

For an example.

jszwedko reviewed (commented):
Thanks for this contribution @r3b-fish !

In general it looks good to me. I just think we should enforce the allowed methods at deserialization time so users get an error instead of it silently falling back to `POST`. See below comment.

Alternatively, we could avoid using the `warp` method filters and just use `warp::filters::method::method` to do the check ourselves and let users use whatever HTTP method they want (even one unspecified in https://datatracker.ietf.org/doc/html/rfc7231#section-4.1).

r3b-fish commented:
> Thanks for this contribution @r3b-fish !
> 
> In general it looks good to me. I just think we should enforce the allowed methods at deserialization time so users get an error instead of it silently falling back to `POST`. See below comment.
> 
> Alternatively, we could avoid using the `warp` method filters and just use `warp::filters::method::method` to do the check ourselves and let users use whatever HTTP method they want (even one unspecified in https://datatracker.ietf.org/doc/html/rfc7231#section-4.1).

Thanks for your comments. 
Also, I've checked `warp::filters::method::method` and I think, that this method should use to extract `Method` from route. May you add some an example for me? 

Many thanks

jszwedko commented on src/sources/util/http/prelude.rs:
```diff
@@ -45,6 +47,7 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
         self,
         address: SocketAddr,
         path: &str,
+        http_method: Method,
```
I would have this take the `HttpMethod` type you defined above. That way the below `match` could be an exhaustive match on the enum rather than following back to `POST` on line 71 (which would never actually be hit in practice since we verify earlier that the incoming method is one of a defined set).

r3b-fish commented:
@jszwedko I've removed `Option`. Also, I've renamed `http_method` in `method`, because sinks/http has the same option. Can you check it please, thanks.

jszwedko reviewed (approved):
Awesome, good work. Thanks for this contribution @r3b-fish !

r3b-fish commented:
@jszwedko  I've restored Option<HttpMethod> because it needs to keep backward compatibility. You can check e2e tests to make sure. Can you review my code. Thanks.

jszwedko reviewed (commented):
Hi @r3b-fish ! Can you clarify what the compatibility issue was? You should be able to avoid using `Option` and use `#serde[default]` on the field (you can see some other examples of that on the config fields).

r3b-fish commented:
Hi @jszwedko! Yes, my previous version of code breaks [e2e test](https://github.com/vectordotdev/vector/blob/master/tests/end-to-end.rs#L74) because `method` should be specified for simple config.

jszwedko commented:
Hi @r3b-fish !

I was able to remove the `Option` and still have the test pass with this diff:

```diff
544ed3644 - (HEAD -> feature/add-support-for-http-methods, origin/feature/add-support-for-http-methods) Remove Option (2 minutes ago) <Jesse Szwedko>
diff --git a/src/sources/http.rs b/src/sources/http.rs
index 15b7c92a5..9b7a39ca6 100644
--- a/src/sources/http.rs
+++ b/src/sources/http.rs
@@ -56,7 +56,8 @@ pub(super) struct SimpleHttpConfig {
     path: String,
     #[serde(default = "default_path_key")]
     path_key: String,
-    method: Option<HttpMethod>,
+    #[serde(default)]
+    method: HttpMethod,
     framing: Option<FramingConfig>,
     decoding: Option<DeserializerConfig>,
     #[serde(default, deserialize_with = "bool_or_struct")]
@@ -78,7 +79,7 @@ impl GenerateConfig for SimpleHttpConfig {
             auth: None,
             path: "/".to_string(),
             path_key: "path".to_string(),
-            method: Some(HttpMethod::Post),
+            method: HttpMethod::Post,
             strict_path: true,
             framing: None,
             decoding: Some(default_decoding()),
@@ -96,10 +97,6 @@ fn default_path_key() -> String {
     "path".to_string()
 }
 
-const fn default_http_method() -> HttpMethod {
-    HttpMethod::Post
-}
-
 #[derive(Clone)]
 struct SimpleHttpSource {
     headers: Vec<String>,
@@ -195,15 +192,11 @@ impl SourceConfig for SimpleHttpConfig {
             path_key: self.path_key.clone(),
             decoder,
         };
-        let http_method = match self.method.as_ref() {
-            Some(http_method) => *http_method,
-            None => default_http_method(),
-        };
 
         source.run(
             self.address,
             self.path.as_str(),
-            http_method,
+            self.method,
             self.strict_path,
             &self.tls,
             &self.auth,
@@ -322,7 +315,7 @@ mod tests {
                 strict_path,
                 path_key,
                 path,
-                method: Some(method),
+                method,
                 framing,
                 decoding,
                 acknowledgements: acknowledgements.into(),
```

I tried to push it to this branch, but I think edits by maintainers might be disabled. Would you mind applying this diff? I think this looks good otherwise!

r3b-fish commented:
@jszwedko updated

jszwedko commented:
Great, thanks @r3b-fish ! I marked this to merge when CI passes. Thanks again for this contribution!

r3b-fish commented:
I don't know that we can use `#[serde(default)]`. It's my first experience to write on Rust. Vector is a cool project.
Thanks @jszwedko that you helped me with it

r3b-fish commented:
@jszwedko I've rebased branch to resolve merge conflicts. Can you restart CI-checks, please. Thanks!

jszwedko commented:
Thanks again for this contribution @r3b-fish !
